core/profile.c

```diff
@@ -1255,14 +1255,14 @@ static void calculate_gas_information_new(const struct dive *dive, const struct
 		 * END takes O₂ + N₂ (air) into account ("Narcotic" for trimix dives)
 		 * EAD just uses N₂ ("Air" for nitrox dives) */
 		pressure_t modpO2 = { .mbar = (int)(prefs.modpO2 * 1000) };
-		entry->mod = (double)gas_mod(gasmix, modpO2, dive, 1).mm;
-		entry->end = (entry->depth + 10000) * (1000 - fhe) / 1000.0 - 10000;
-		entry->ead = (entry->depth + 10000) * fn2 / (double)N2_IN_AIR - 10000;
-		entry->eadd = (entry->depth + 10000) *
+		entry->mod = gas_mod(gasmix, modpO2, dive, 1).mm;
+		entry->end = mbar_to_depth((int)(depth_to_mbarf(entry->depth, dive) * (1000 - fhe) / 1000.0), dive);
```

Comment:
This code is mathematically wrong.

`(int)(...)` will round 100.0-100.0*eps (99.9999999...) to 99.

You want `lrint(...)` or `(int)(round(...))`, as I have stated before.

Of course, that will still give the same rounding problem just for 100.5 +/0 100.5 * eps. But at least the result is not outright wrong.

To fully avoid the problem, I reckon `mbar_to_depth` should also exist as version taking a double.

Reply:
You are right, the (int) cast always rounds down not to nearest value. 

I don't think the problem at 100.5 is as big as we have many situations where we are at an int (e.g. END for nitrox dives) then do calculation that for real numbers is a NOOP but not for floating point and for (int) we were doing that at the edge of the discontinuity. But I would expect much fewer cases where we are sitting close the the .5 edge.

I disagree, though that a double version of mbar_to_depth would help: Then, in the test, we would be comparing doubles for equality which is bad, comparing the rounded ints has much better chance of not picking up noise.

Reply:
> I disagree, though that a double version of mbar_to_depth would help: Then, in the test, we would be comparing doubles for equality which is bad, comparing the rounded ints has much better chance of not picking up noise.

You misunderstood. I'm not suggesting to return a double, but to take a double argument. The fundamental problem in all this code is intermittent conversion to integral types. Fixed-point and floating-point are both fine, if done correctly. It's the combination that ends up in a disaster.